src/addr_validate.rs

```diff
@@ -0,0 +1,91 @@
+use std::{cell::RefCell, mem::size_of};
+
+use nix::{
+    errno::Errno,
+    fcntl::OFlag,
+    unistd::{close, pipe2, read, write},
+};
+
+thread_local! {
+    static MEM_VALIDATE_PIPE: RefCell<[i32; 2]> = RefCell::new([-1, -1]);
+}
+
+fn open_pipe() -> nix::Result<()> {
+    MEM_VALIDATE_PIPE.with(|pipes| {
+        let mut pipes = pipes.borrow_mut();
+
+        // ignore the result
+        let _ = close(pipes[0]);
+        let _ = close(pipes[1]);
+
+        let (read_fd, write_fd) = pipe2(OFlag::O_CLOEXEC | OFlag::O_NONBLOCK)?;
+
+        pipes[0] = read_fd;
+        pipes[1] = write_fd;
+
+        Ok(())
+    })
+}
+
+pub fn validate(addr: *const libc::c_void) -> bool {
+    const CHECK_LENGTH: usize = 2 * size_of::<*const libc::c_void>() / size_of::<u8>();
+
+    // read data in the pipe
+    let valid_read = MEM_VALIDATE_PIPE.with(|pipes| {
+        let pipes = pipes.borrow();
+        loop {
+            let mut buf = [0u8; CHECK_LENGTH];
+
+            match read(pipes[0], &mut buf) {
+                Ok(bytes) => break bytes > 0,
+                Err(_err @ Errno::EINTR) => continue,
+                Err(_err @ Errno::EAGAIN) => break true,
+                Err(_) => break false,
+            }
+        }
+    });
```

Comment:
Why do we need to try to read `pipes[0]` first? Can we just try to write `pipes[1]` directly?

Reply:
Fine, i see, this is to clear the previously written data...